staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go

```diff
@@ -0,0 +1,142 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+	"strings"
+)
+
+var (
+	// health probes and metrics scraping are never rejected, we will continue
+	// serving these requests after shutdown delay duration elapses.
+	excludedPathPrefixes = []string{
+		"/readyz",
+		"/livez",
+		"/healthz",
+		"/metrics",
+	}
+)
+
+// excludeFunc returns true if the request should be excluded,
+// otherwise it returns false.
+// excluded requests are always served.
+type excludeFunc func(*http.Request) bool
+
+// retryAfterParams dictates how the Retry-After response is constructed
+type retryAfterParams struct {
+	// TearDownConnection is true when we should send a
+	// 'Connection: close' header in the response.
+	TearDownConnection bool
+
+	// Message describes why Retry-After response has been sent by the server
+	Message string
+}
+
+// retryAfterConditionFunc returns true if the requests should
+// be rejected once certain conditions are met.
+// The retryAfterParams returned contains instructions on how to
+// construct the Retry-After response.
+type retryAfterConditionFunc func() (*retryAfterParams, bool)
+
+func (f retryAfterConditionFunc) ShouldSendRetryAfterResponse() (*retryAfterParams, bool) {
+	return f()
+}
+
+// WithRetryAfter rejects any incoming new request(s) with a 429
+// if the specified shutdownDelayDurationElapsedFn channel is closed
+//
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving after shutdownDelayDurationElapsedFn is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance)
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+	return withRetryAfter(handler, func(r *http.Request) bool {
+		if hasExcludedPathPrefix(r) {
+			return true
+		}
+		if isKubeApiserverUserAgent(r) {
+			return true
+		}
+		return false
+	}, newRetryAfterFunc(shutdownDelayDurationElapsedCh))
+}
+
+func withRetryAfter(handler http.Handler, excludeFn excludeFunc, retryAfterFn retryAfterConditionFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := retryAfterFn.ShouldSendRetryAfterResponse()
+		if !sendRetryAfter {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// if we are here this means it's time to send Retry-After response to
+		// all requests except the ones that are excluded.
+		if excludeFn(req) {
+			handler.ServeHTTP(w, req)
```

Comment:
I think that it will keep the `drained` queue busy and will block the server from shutting down, am I right?

Reply:
we can exlucde them from the wait group too

Reply:
yeah, for now using the same exclude function in `WithWaitGroup`, I have left a todo to merge these two filters in the future.